Refactor postshader handling to apply for softgpu / D3D9

unknownbrackets opened the pull request:
This splits out the FramebufferManagerCommon class, creating a PresentationCommon to handle things like post-processing shaders, Google Cardboard, and anything else with the final display.

With that split out, it's now used everywhere and more consistent.  Benefits:
 * Software renderer now supports Google Cardboard and post-processing shaders (fixes #12445.)
 * Google Cardboard now supports render rotation.
 * Direct3D 9 now supports (some) post-processing shaders.  All ours work if you have PS 3.0 hardware.
 * There are a bit less differences between backends in post-processing shader handling, so features like chaining shaders will be easier to add now.

Also some cleanup and code centralization.  

Some remaining concerns:
 * No idea if this actually works with DarkStalkers.
 * Flipped buffers are still a bit messy.
 * Some minor optimizations might need reapplying (currently does 2x triangles rather than a strip, for example.)
 * OpenGL (non-Core) and GLES may not report errors as well as before.
 * Vulkan now uses the thin3d texture allocator for draw pixels...

-[Unknown]

unknownbrackets committed:
softgpu: Fix texture overlap.

Mainly happened when we had wide textures and split them up between GPUs.

unknownbrackets committed:
GPU: Split FramebufferCommon into two classes.

Only some things moved over so far.

FramebufferCommon does too much, we want to share it with softgpu without
all the buffer management stuff.

unknownbrackets committed:
Vulkan: Fix Draw usage of framebuffers.

unknownbrackets committed:
GPU: Move cardboard/etc. to PresentationCommon.

Now this works on softgpu as well.

Some hacks for backend differences...

unknownbrackets committed:
softgpu: Allow configuring Google Cardboard.

This is also relevant where softgpu is forced, i.e. Darkstalkers.

unknownbrackets committed:
softgpu: Allow display rotation.

unknownbrackets committed:
GPU: Move a bit more to new display approach.

unknownbrackets committed:
Draw: Allow specifying a tag on shaders.

This way they are not always "thin3d" when using Draw.

unknownbrackets committed:
Draw: Add more uniform types.

Only because we might have float a[2]; float b[2];...

unknownbrackets committed:
Draw: Enable backwards compat on D3D11 shaders.

To match/support postshaders.

unknownbrackets committed:
D3D9: Enable postprocessing of post shaders.

unknownbrackets committed:
GPU: Move post shader handling to new class.

Currently, Vulkan is not working properly and direct (RAM -> output) is
not hooked up.  But in general, it works.

unknownbrackets committed:
UI: Allow postshaders on D3D9.

Since they work now - at least, some of the shaders do.

unknownbrackets committed:
GPU: Cleanup leftover postshader stuff.

unknownbrackets committed:
GLES: Fix reuse of vertex data buffer.

Direct3D 11 and OpenGL had some issues with this being reused for both
renders.  Caused OpenGL to appear flipped.

unknownbrackets committed:
softgpu: Enable postshader support.

unknownbrackets committed:
Draw: Actually use the index offset parameter.

I'm not sure we need it, but having a param used only on one backend is
confusing.

unknownbrackets committed:
Vulkan: Correct postshader sampler binding.

This was causing a lost device.

unknownbrackets committed:
Vulkan: Correct missing offsets in Draw.

Was silently ignoring them.  Caused stretch in postshaders.

unknownbrackets committed:
Vulkan: Allow frag shader UBO reads.

unknownbrackets committed:
softgpu: Avoid RB swizzle when using a postshader.

So that it can post-process correctly.

LunaMoo commented:
Doesn't work with DarkStalkers even after disabling it's compat hacks, through it does work with some other games with software renderer.

unknownbrackets committed:
GPU: Use a texture directly for MakePixelTexture.

This makes it easier to do things with it.

unknownbrackets committed:
GPU: Use postshader for direct VRAM draws again.

Also centralize the pixel texture code while at it.

LunaMoo commented:
As for shaders failing in D3d9 I tried to figure out what fails as the error is very generic and only says it failed to build and it seems that calling texture2D with texcoord changed by different values more than 9 times or so are causing the failure. If I try to call the function many times, but with same value, it doesn't fail, so I'm guessing the compiler optimizes that situation, but it can't optimize the calls to different coordinates. Not sure how, but maybe this could be detected and instead of generic failure error, show something like "This effect doesn't work under D3D9"?

unknownbrackets commented:
Yeah, we indeed use ps 2.0:

```
const char *profile = stage_ == ShaderStage::FRAGMENT ? "ps_2_0" : "vs_2_0";
```

Updated this with some of the things implemented, like the pixel changes.  Now it should work with homebrew etc. and the RB swizzle thing might be better.

-[Unknown]

unknownbrackets committed:
D3D9: Attempt PS/VS 3.0 if it doesn't work with 2.

So that postshaders can work.  Downside is, if we write a UI shader that
requires 2.0, we may not notice... but I think the risk is not that high.

LunaMoo commented:
In Darkstalkers menu using post process shaders now causes:
![Clipboard02](https://user-images.githubusercontent.com/5485237/81624227-f4b51480-93f5-11ea-893f-1e136035e60f.jpg)
 and in-game it's just black screen. Previously it just wasn't showing any change. Dump can reproduce it when using software renderer:
[ULUS10005_0001.zip](https://github.com/hrydgard/ppsspp/files/4612736/ULUS10005_0001.zip)

 Also in other games using 5xBR shows artifacts in software renderer, kind of like the coords, pixel or texel size was wrong or at least less accurate.

unknownbrackets committed:
softgpu: Fix postshader on 5551.

This also fixes rendering on Windows 7 Direct3D 11.

unknownbrackets commented:
Ah, the no-16 bit support path was just broken, even before these changes.  It should work now at least for the menu.

-[Unknown]

unknownbrackets committed:
GPU: Simplify direct render when skipping buffers.

This may have been uploading twice before, so there may be a small
performance improvement in some games now.

unknownbrackets committed:
GPU: Cleanup skip buffer texture draws.

These will be used when a game does multiple transfers from RAM to the
screen, rather than one big one.  Wasn't clearing some state, though.

LunaMoo commented:
In-game still black screen and on the boot screen it shows enlarged upper left of the screen blinking, unfortunately non of that shows in dumps, but I noticed it's caused by the (speed)hack used for this game, without it, it works correctly, in fact better than in other games now as it's accurate there:).

 The lack of accuracy in other games persist through, try using any 5xBR variant on this dump with software renderer:
[ULES00469_0001.zip](https://github.com/hrydgard/ppsspp/files/4613205/ULES00469_0001.zip)

unknownbrackets commented:
Hm, I get artifacts even on OpenGL (non-software) with that dump.  Also, for some reason Direct3D 11 is not showing anything.. hmm.

The formula for pixelDelta in softgpu for that dump is:
```
u_pixelDelta.x = 1/w * 480/bufferWidth
w = 1280.000000 # Was showing this at 1280x1080
1/w = 0.000781
bufferWidth = 512 (size of texture)
480/bufferWidth = 0.9375
--
u_pixelDelta.x  = 0.000781 * 0.9375 = 0.0007321875
```

But, that seems wrong?  What does the buffer (texture) width have to do with pixelDelta?

In hardware mode, fb textures will usually but not always be a multiple of 480 wide, and bufferWidth will almost always be 480.  Maybe this just wasn't noticed, but was always wrong?  It looks like hardware if I set it to 480.

But #8016 says I'm wrong and that this is very intentional.  The buffer is definitely 512 wide, though... note that the 481 thing is fixed separately.

-[Unknown]

unknownbrackets committed:
D3D11: Allow pixel shader uniforms in Draw.

unknownbrackets committed:
GPU: Cleanup presentation flipping a bit.

LunaMoo commented:
I can improve the xBR in software renderer by changing the one u_pixelDelta  to u_texelDelta, this will break it for hardware renderers so not sure what happens there.

 [I also have a version which uses only u_pixelDelta](https://forums.ppsspp.org/showthread.php?tid=193&pid=134556#pid134556) to work in opposite and much more wasteful way to make use of the increased rendering res as well as support custom stretch values to perfectly match ports from older consoles which had different pixel sizes and a lot of blur introduced, that seems to work fine, but it was made for SFA3Max weird pixel sizes and needs to be tweaked.

hrydgard commented:
Whoa this change grew :) 

I'll try to review it tomorrow but looking very promising. The u_pixelDelta thing looks suspicious indeed.

unknownbrackets commented:
Heh, sorry.  If needed, I can break out the Draw changes and maybe some of the non-postshader stuff.

I haven't written any post shaders, but from the names, here's what I'd imagine those variables to do:

Name | Imagination | Actual | Actual @ output res
--- | --- | ---
pixelDelta | Value to add to gl_Position to get to next dest pixel, i.e. `2.0 / targetSize` | Half this value | Half this value scaled down by the padding in the texture
texelDelta | Value to add to texture2D(param) to get the next texel, i.e. `1.0 / bufferSize` | 1.0 / renderResolutionSize (buffer may not be 480 * res) | Same
time | Values that increase each frame | Same | Same
video | 1 if there are videos | Same | Same

I suspect maybe texelDelta is wrong, and this motivated us to make pixelDelta also wrong to compensate?

-[Unknown]

unknownbrackets commented:
See d6d1fe681 - this makes the uniforms match my expectation.  I think it should merge separately, because the current behavior is in line with what it's been doing.  I don't know what this would break...

-[Unknown]

unknownbrackets committed:
GPU: Centralize DestroyAllFBOs().

This cleans up the postshader update code.

hrydgard commented:
Looks right - we could merge that one first actually.

unknownbrackets committed:
GPU: Use texclamp for postshader/present draw.

hrydgard commented:
There's a conflict, but tried building directly from your branch and it crashes in GPUCommon::ReInitialize during SoftGPU::SoftGPU, calling textureCache_->Clear(true);   We need to initialize textureCache_ to nullptr, and check for it here to make sure we don't do anything with it if it's not set. Dunno if that really was introduced here though...

Anyway, that should be trivial, and I'll look through code now.

hrydgard commented on GPU/Common/FramebufferCommon.cpp:
```diff
@@ -778,56 +702,99 @@ void FramebufferManagerCommon::CopyFramebufferForColorTexture(VirtualFramebuffer
+				if (preferredPixelsFormat_ == Draw::DataFormat::B8G8R8A8_UNORM)
+					ConvertRGB565ToBGRA8888(dst, src16, width);
+				else
+					ConvertRGBA565ToRGBA8888(dst, src16, width);
```
Heh, just noticed, we should rename ConvertRGBA565ToRGBA8888 to ConvertRGB565ToRGBA8888  (removed an A).

hrydgard commented on GPU/Common/FramebufferCommon.cpp:
```diff
@@ -778,56 +702,99 @@ void FramebufferManagerCommon::CopyFramebufferForColorTexture(VirtualFramebuffer
-		SetViewport2D(0, 0, pixelWidth_, pixelHeight_);
-		DrawActiveTexture(x, y, w, h, (float)pixelWidth_, (float)pixelHeight_, u0, v0, u1, v1, uvRotation, flags);
+	// DrawActiveTexture reverses these, probably to match "up".
+	if (GetGPUBackend() == GPUBackend::DIRECT3D9 || GetGPUBackend() == GPUBackend::DIRECT3D11) {
```
I'll clean up this later by fixing the shaders.

hrydgard reviewed (commented):
Really a pretty great refactoring, and nice fixes to thin3d stuff!

hrydgard commented:
DarkStalkers is not working as-is (glitches, blackness) but I'll help fix that up once this is ready (which it very nearly is).

unknownbrackets committed:
GPU: Take A off RGB565 conversion funcs.

unknownbrackets commented:
Rebased - yeah, I must've broke softgpu in centralizing DestroyAllFBOs(), oops.  Fixed now.

Not sure what's happening in Darkstalkers - it should be right, unless something is wrong with the UVs?

-[Unknown]

LunaMoo commented:
DarkStalkersPresentHack is causing the failure, it works without it just fine.

 Althrough when I disabled it and tried to take a screenshot I noticed one other bad thing about software renderer - the screenshots will always be x1 res, so if I want to show results of upscaling shader I'd have to take screenshot externally without using the built-in function.

unknownbrackets commented:
I know, but outside the smaller UVs, that code path is very similar...

-[Unknown]

unknownbrackets committed:
Draw: Small optimization to callback.

This allows the callback to say "I didn't copy, please copy for me."  This
helps when additional conversions will be applied during the copy, or a
copy can be skipped.

Also, fixed a couple cases of the wrong mip level being used.

hrydgard commented:
I'm not sure either, but I'll look into it tonight. It looks like it should work.

LunaMoo commented:
Weird thing is it bugs out even with my empty effect that I used in the past just for adding the 60fps flag. So just setting texture2D(sampler0, v_texcoord0.xy) into output color is enough to break things and it breaks differently on different screens as well.

hrydgard commented:
Alright, I found the problem. In DarkStalkers we directly blit from a specific address where the game draws the buffer to be displayed. The function ConvertTextureDescFrom16 did not take that into account but directly used displayFramebuf_, so added a parameter to override. Will push that after merging - which I will do now.

hrydgard commented:
The fix commit, for reference: https://github.com/hrydgard/ppsspp/commit/864d138cd95f932b91ec7a7bc738f1eaca627667
